parser-typechecker/src/Unison/Codebase/Init.hs

```diff
@@ -47,67 +47,61 @@ type DebugName = String
 
 data Init m v a = Init
   { -- | open an existing codebase
-    openCodebase :: DebugName -> CodebasePath -> m (Either Pretty (m (), Codebase m v a)),
+    withOpenCodebase :: forall r. DebugName -> CodebasePath -> (Codebase m v a -> m r) -> m (Either Pretty r),
     -- | create a new codebase
-    createCodebase' :: DebugName -> CodebasePath -> m (Either CreateCodebaseError (m (), Codebase m v a)),
+    withCreatedCodebase :: forall r. DebugName -> CodebasePath -> (Codebase m v a -> m r) -> m (Either CreateCodebaseError r),
     -- | given a codebase root, and given that the codebase root may have other junk in it,
     -- give the path to the "actual" files; e.g. what a forked transcript should clone.
     codebasePath :: CodebasePath -> CodebasePath
   }
 
-type FinalizerAndCodebase m v a = (m (), Codebase m v a) 
-
-data InitError 
+data InitError
   = NoCodebaseFoundAtSpecifiedDir
   | FoundV1Codebase
   | CouldntCreateCodebase Pretty
 
-data InitResult m v a 
-  = OpenedCodebase CodebasePath (FinalizerAndCodebase m v a) 
-  | CreatedCodebase CodebasePath (FinalizerAndCodebase m v a) 
-  | Error CodebasePath InitError 
```

Comment:
Since the codebase is passed to the action, but the error means that the action isn't ever called, I split off the error case so things make more sense.